src/Bicep.Core/Workspaces/SourceFileGroupingBuilder.cs

```diff
@@ -161,291 +161,165 @@ private ResultWithDiagnostic<ISourceFile> PopulateRecursive(Uri fileUri, Artifac
             {
-                var (childArtifactReference, uriResult) = GetArtifactRestoreResult(file.FileUri, restorable);
-                if (!uriResultByBicepSourceFileByArtifactReferenceSyntax.TryGetValue(file, out var uriResultByArtifactReferenceSyntaxLookup))
+                if (restorable is ProviderDeclarationSyntax providerDeclaration)
                 {
-                    uriResultByArtifactReferenceSyntaxLookup = [];
-                    uriResultByBicepSourceFileByArtifactReferenceSyntax[file] = uriResultByArtifactReferenceSyntaxLookup;
+                    var isBuiltInProvider = providerDeclaration.Specification switch {
+                        LegacyProviderSpecification => true,
+                        ConfigurationManagedProviderSpecification configSpec => config.ProvidersConfig.IsSysOrBuiltIn(configSpec.NamespaceIdentifier),
+                        _ => false,
+                    };
+
+                    if (isBuiltInProvider)
+                    {
+                        // built-in provider - no restoration required
+                        continue;
+                    }
+
+                    artifactLookup[restorable] = GetArtifactRestoreResult(file, restorable);
+
+                    // recursion not needed for provider declarations
+                    continue;
                 }
-                uriResultByArtifactReferenceSyntaxLookup[restorable] = uriResult;
 
-                if (!uriResult.IsSuccess(out var artifactUri))
+                var resolutionInfo = GetArtifactRestoreResult(file, restorable);
+                artifactLookup[restorable] = resolutionInfo;
+                if (!resolutionInfo.Result.IsSuccess(out var artifactUri))
                 {
+                    // recursion not possible
                     continue;
                 }
 
                 if (!fileResultByUri.TryGetValue(artifactUri, out var childResult) ||
                     (childResult.IsSuccess(out var childFile) && sourceFilesToRebuild is not null && sourceFilesToRebuild.Contains(childFile)))
                 {
                     // only recurse if we've not seen this file before - to avoid infinite loops
-                    childResult = PopulateRecursive(artifactUri, childArtifactReference, sourceFilesToRebuild, featureProviderFactory, configurationManager);
+                    childResult = PopulateRecursive(artifactUri, resolutionInfo.Reference, sourceFilesToRebuild, featureProviderFactory, configurationManager);
                 }
                 fileResultByUri[artifactUri] = childResult;
             }
+
+            foreach (var providerName in config.ImplicitProvidersConfig.GetImplicitProviderNames())
+            {
+                if (TryGetImplicitProvider(providerName, file, config) is { } implicitProvider)
+                {
+                    implicitArtifacts.Add(implicitProvider);
+                }
+            }
```

Comment:
nit: It may be easier on the reader to extract this to a separate function and invoke it above the processing of the provider declaration loop, this way its easier to follow that we process all providers, and once done we go ahead and process other artifacts.